source/traction_boundary_conditions/lithostatic_pressure.cc

```diff
@@ -0,0 +1,469 @@
+                                                       + dealii::Utilities::int_to_string(i)
+                                                       + std::string(" is bigger than the size of the pressure vector ")
+                                                       + dealii::Utilities::int_to_string(pressure.size())));
+
+          // current depth z
+          z = double(i) * delta_z;
+
+          // set up the input for the density function of the material model
+          typename MaterialModel::Interface<dim>::MaterialModelInputs in(1, n_compositional_fields);
+          typename MaterialModel::Interface<dim>::MaterialModelOutputs out(1, n_compositional_fields);
+
+          // where to calculate the density
+          // for spherical domains
+          if (dynamic_cast<const GeometryModel::SphericalShell<dim>*> (&this->get_geometry_model()) != 0 ||
+              dynamic_cast<const GeometryModel::Sphere<dim>*> (&this->get_geometry_model()) != 0 ||
+              dynamic_cast<const GeometryModel::Chunk<dim>*> (&this->get_geometry_model()) != 0 ||
+              /*dynamic_cast<const GeometryModel::TwoMergedChunks<dim>*> (&this->get_geometry_model()) != 0*/
+              dynamic_cast<const GeometryModel::EllipsoidalChunk<dim>*> (&this->get_geometry_model()) != 0)
+            {
+              // decrease radius with depth increment
+              spherical_representative_point[0] -= delta_z;
+              in.position[0] = spherical_to_cart(spherical_representative_point);
+            }
+          // and for cartesian domains
+          else
+            {
+              // decrease z coordinate with depth increment
+              representative_point[dim-1] -= delta_z;
+              in.position[0] = representative_point;
+            }
+
+          // we need the actual temperature at t0
+          in.temperature[0] = this->get_initial_conditions().initial_temperature(in.position[0]);
+          // the previous pressure
+          in.pressure[0] = pressure[i-1];
+
+          // we need the actual composition at t0
+          for (unsigned int c=0; c<n_compositional_fields; ++c)
+            in.composition[0][c] = this->get_compositional_initial_conditions().initial_composition(in.position[0], c);
+
+          in.strain_rate.resize(0); // we do not need the viscosity
+          this->get_material_model().evaluate(in, out);
+
+          // get the magnitude of gravity. we assume
+          // that gravity always points along the depth direction. this
+          // may not strictly be true always but is likely a good enough
+          // approximation here.
+          const double density = out.densities[0];
+          const double gravity = this->get_gravity_model().gravity_vector(in.position[0]).norm();
+
+          // trapezoid integration
+          pressure[i] = sum + delta_z * 0.5 * density * gravity;
+          sum += delta_z * density * gravity;
+
```

Comment:
It took me a while to figure out how the integration works. If you only want to do this once at the initialization, wouldn't it be easier to take the simpler logic from the forward Euler (instead of the trapezoidal rule) and just take more points? Or have you tried that and it didn't work?